tracer/src/Datadog.Trace/Headers/Ip/RequestIpExtractor.cs

```diff
@@ -0,0 +1,72 @@
+// <copyright file="RequestIpExtractor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.Headers.Ip
+{
+    internal static class RequestIpExtractor
+    {
+        private static readonly IReadOnlyList<string> IpHeaders = new[] { "x-forwarded-for", "x-real-ip", "client-ip", "x-forwarded", "x-cluster-client-ip", "forwarded-for", "forwarded", "via", "true-client-ip" };
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(RequestIpExtractor));
+
+        internal static IpInfo ExtractIpAndPort(Func<string, string> getHeader, string customIpHeader, bool isSecureConnection, IpInfo peerIpFallback)
+        {
+            IpInfo extractedCustomIp = null;
+            if (!string.IsNullOrEmpty(customIpHeader))
+            {
+                var value = getHeader(customIpHeader);
+                if (!string.IsNullOrEmpty(value))
+                {
+                    extractedCustomIp = IpExtractor.RealIpFromValue(value, isSecureConnection);
+                    if (extractedCustomIp == null)
+                    {
+                        Log.Warning("A custom header for ip with value {value} was configured but no ip could be read", value);
+                        return null;
+                    }
+
+                    // dont return extractedCustomIp if not null yet, we want to check that no other ip headers are found, otherwise could be a sign of injection, i.e. RFC on ip extraction algorithm. If other headers are to be found, then no result should be returned
```

Comment:
Based on this text from the RFC, it looks like we should return early here with the extracted custom IP. I'm not sure if it got updated since you wrote this or if the RFC contradicts itself somewhere:


> The client may set the DD_TRACE_CLIENT_IP_HEADER environment variable indicating a header for collecting IP. This should contain the name of a single header, configuring multiple headers is not allowed. If this variable is set all other IP related headers should be ignored.